### PR TITLE
Basic interactions with vertices

### PR DESCRIPTION
**Pull request by StefanPaulet**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -36,27 +36,28 @@ else()
 endif()
 
 set(
-    AGE_CORE_INCLUDE_DIRECTORIES
-    ${CMAKE_SOURCE_DIR}/dependencies/cds/include
-    ${CMAKE_SOURCE_DIR}/src/core
+        AGE_CORE_INCLUDE_DIRECTORIES
+        ${CMAKE_SOURCE_DIR}/dependencies/cds/include
+        ${CMAKE_SOURCE_DIR}/src/core
```
Comment (LoghinVladDev) [inline CMakeLists.txt]: I know this is a pain - 2 spaces in CMakeLists.txt as well. Adjust your auto code formatting in your IDE to match this. 
```suggestion
    AGE_CORE_INCLUDE_DIRECTORIES
    ${CMAKE_SOURCE_DIR}/dependencies/cds/include
    ${CMAKE_SOURCE_DIR}/src/core
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -36,27 +36,28 @@ else()
...
 )
 
 set(
-    CORE_SOURCES
-    ${CMAKE_SOURCE_DIR}/src/core/lang/string/StringRef.cpp
-    ${CMAKE_SOURCE_DIR}/src/core/lang/filesystem/PathAwareFstream.cpp
+        CORE_SOURCES
+        ${CMAKE_SOURCE_DIR}/src/core/lang/string/StringRef.cpp
+        ${CMAKE_SOURCE_DIR}/src/core/lang/filesystem/PathAwareFstream.cpp
+        src/core/intern/QtDefines.hpp
```
Comment (LoghinVladDev) [inline CMakeLists.txt]: Same comment from l:39-l:41
```suggestion
    CORE_SOURCES
    ${CMAKE_SOURCE_DIR}/src/core/lang/string/StringRef.cpp
    ${CMAKE_SOURCE_DIR}/src/core/lang/filesystem/PathAwareFstream.cpp
    src/core/intern/QtDefines.hpp
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -36,27 +36,28 @@ else()
...
 )
 
 add_library(
-    lib.core
-    STATIC
-    ${CORE_SOURCES}
+        lib.core
+        STATIC
+        ${CORE_SOURCES}
```
Comment (LoghinVladDev) [inline CMakeLists.txt]: Same comment from l:39-l:41
```suggestion
    lib.core
    STATIC
    ${CORE_SOURCES}
```

File: CMakeLists.txt [hunk 4]
```diff
@@ -36,27 +36,28 @@ else()
...
 )
 
 target_include_directories(
-    lib.core
-    PUBLIC
-    ${AGE_CORE_INCLUDE_DIRECTORIES}
+        lib.core
+        PUBLIC
+        ${AGE_CORE_INCLUDE_DIRECTORIES}
```
Comment (LoghinVladDev) [inline CMakeLists.txt]: Same comment from l:39-l:41
```suggestion
    lib.core
    PUBLIC
    ${AGE_CORE_INCLUDE_DIRECTORIES}
```

File: CMakeLists.txt [hunk 5]
```diff
@@ -66,61 +67,62 @@ if(DEFINED QT_VERSION)
   set(CMAKE_AUTOUIC ON)
 
   set(
-      VISUALIZER_CORE_SOURCES
-      ${CMAKE_SOURCE_DIR}/src/visualizer/settings/SettingsRegistry.cpp
+          VISUALIZER_CORE_SOURCES
+          ${CMAKE_SOURCE_DIR}/src/visualizer/settings/SettingsRegistry.cpp
```
Comment (LoghinVladDev) [inline CMakeLists.txt]: Same comment from l:39-l:41
```suggestion
     VISUALIZER_CORE_SOURCES
     ${CMAKE_SOURCE_DIR}/src/visualizer/settings/SettingsRegistry.cpp
```

File: CMakeLists.txt [hunk 6]
```diff
@@ -66,61 +67,62 @@ if(DEFINED QT_VERSION)
...
   set(
-      VISUALIZER_SOURCES
-      ${CMAKE_SOURCE_DIR}/src/visualizer/window/VisualizerWindow.cpp
-      ${CMAKE_SOURCE_DIR}/src/visualizer/vertex/Vertex.cpp
-      ${CMAKE_SOURCE_DIR}/src/visualizer/graphPanel/GraphPanel.cpp
+          VISUALIZER_SOURCES
+          ${CMAKE_SOURCE_DIR}/src/visualizer/window/VisualizerWindow.cpp
+          ${CMAKE_SOURCE_DIR}/src/visualizer/vertex/Vertex.cpp
+          ${CMAKE_SOURCE_DIR}/src/visualizer/vertexMenu/VertexMenu.cpp
+          ${CMAKE_SOURCE_DIR}/src/visualizer/graphPanel/GraphPanel.cpp
```
Comment (LoghinVladDev) [inline CMakeLists.txt]: Same comment from l:39-l:41
```suggestion
     VISUALIZER_SOURCES
     ${CMAKE_SOURCE_DIR}/src/visualizer/window/VisualizerWindow.cpp
     ${CMAKE_SOURCE_DIR}/src/visualizer/vertex/Vertex.cpp
     ${CMAKE_SOURCE_DIR}/src/visualizer/vertexMenu/VertexMenu.cpp
     ${CMAKE_SOURCE_DIR}/src/visualizer/graphPanel/GraphPanel.cpp
```

File: CMakeLists.txt [hunk 7]
```diff
@@ -66,61 +67,62 @@ if(DEFINED QT_VERSION)
...
   )
 
   add_library(
-      lib.visualizer_core
-      STATIC
-      ${VISUALIZER_CORE_SOURCES}
+          lib.visualizer_core
+          STATIC
+          ${VISUALIZER_CORE_SOURCES}
```
Comment (LoghinVladDev) [inline CMakeLists.txt]: Same comment from l:39-l:41
```suggestion
    lib.visualizer_core
    STATIC
    ${VISUALIZER_CORE_SOURCES}
```

File: CMakeLists.txt [hunk 8]
```diff
@@ -66,61 +67,62 @@ if(DEFINED QT_VERSION)
...
   )
 
   target_link_libraries(
-      lib.visualizer_core
-      PUBLIC
-      lib.core
+          lib.visualizer_core
+          PUBLIC
+          lib.core
```
Comment (LoghinVladDev) [inline CMakeLists.txt]: Same comment from l:39-l:41
```suggestion
    lib.visualizer_core
    PUBLIC
    lib.core
```

File: CMakeLists.txt [hunk 9]
```diff
@@ -66,61 +67,62 @@ if(DEFINED QT_VERSION)
...
   )
 
   add_executable(
-      visualizer
-      ${VISUALIZER_SOURCES}
-      src/target/visualizer.cpp
+          visualizer
+          ${VISUALIZER_SOURCES}
+          src/target/visualizer.cpp
```
Comment (LoghinVladDev) [inline CMakeLists.txt]: Same comment from l:39-l:41
```suggestion
    visualizer
    ${VISUALIZER_SOURCES}
    src/target/visualizer.cpp
```

File: CMakeLists.txt [hunk 10]
```diff
@@ -66,61 +67,62 @@ if(DEFINED QT_VERSION)
...
-      lib.visualizer_core
-      Qt${QT_VERSION}::Core
-      Qt${QT_VERSION}::Gui
-      Qt${QT_VERSION}::Widgets
+          visualizer
+          lib.core
+          lib.visualizer_core
+          Qt${QT_VERSION}::Core
+          Qt${QT_VERSION}::Gui
+          Qt${QT_VERSION}::Widgets
```
Comment (LoghinVladDev) [inline CMakeLists.txt]: Same comment from l:39-l:41
```suggestion
    visualizer
    lib.core
    lib.visualizer_core
    Qt${QT_VERSION}::Core
    Qt${QT_VERSION}::Gui
    Qt${QT_VERSION}::Widgets
```

File: CMakeLists.txt [hunk 11]
```diff
@@ -66,61 +67,62 @@ if(DEFINED QT_VERSION)
...
 
   target_include_directories(
-      visualizer
-      PUBLIC
-      src/visualizer
-      ${AGE_CORE_INCLUDE_DIRECTORIES}
+          visualizer
+          PUBLIC
+          src/visualizer
+          ${AGE_CORE_INCLUDE_DIRECTORIES}
```
Comment (LoghinVladDev) [inline CMakeLists.txt]: Same comment from l:39-l:41
```suggestion
    visualizer
    PUBLIC
    src/visualizer
    ${AGE_CORE_INCLUDE_DIRECTORIES}
```

File: CMakeLists.txt [hunk 12]
```diff
@@ -66,61 +67,62 @@ if(DEFINED QT_VERSION)
...
 
   target_include_directories(
-      lib.visualizer_core
-      PUBLIC
-      src/visualizer
-      ${AGE_CORE_INCLUDE_DIRECTORIES}
+          lib.visualizer_core
+          PUBLIC
+          src/visualizer
+          ${AGE_CORE_INCLUDE_DIRECTORIES}
```
Comment (LoghinVladDev) [inline CMakeLists.txt]: Same comment from l:39-l:41

```suggestion
    lib.visualizer_core
    PUBLIC
    src/visualizer
    ${AGE_CORE_INCLUDE_DIRECTORIES}
```

File: CMakeLists.txt [hunk 13]
```diff
@@ -66,61 +67,62 @@ if(DEFINED QT_VERSION)
...
   )
 endif()
 
 include(FetchContent)
 
 enable_testing()
 add_subdirectory(test/unittests)
-add_subdirectory(test/metatests)
+add_subdirectory(test/metatests)
```
Comment (LoghinVladDev) [inline CMakeLists.txt]: EOL
```suggestion
add_subdirectory(test/metatests)

```

File: src/core/intern/QtDefines.hpp [hunk 14]
```diff
@@ -0,0 +1,9 @@
+//
+// Created by stefan on 8/27/23.
+//
+
+#pragma once
+
+#ifndef AGE_DECLSPEC_SIGNAL
+#define AGE_DECLSPEC_SIGNAL
+#endif
```
Comment (LoghinVladDev) [inline src/core/intern/QtDefines.hpp]: EOL
```suggestion
#endif

```

File: src/visualizer/vertex/Vertex.cpp [hunk 15]
```diff
@@ -4,18 +4,61 @@
...
+                      DRAWN_VERTEX_DIAMETER, DRAWN_VERTEX_DIAMETER);
+}
+
+auto Vertex::mousePressEvent(QMouseEvent* pEvent) -> void {
+  if (distanceToCenter(pEvent->position().toPoint()) > DRAWN_VERTEX_DIAMETER / 2.0f) {
+    return;
+  }
+  switch (pEvent->button()) {
+    case Qt::LeftButton: handleLeftClickEvent(pEvent); break;
+    case Qt::RightButton: emit rightClickPressed(mapToParent(pEvent->position())); break;
```
Comment (LoghinVladDev) [inline src/visualizer/vertex/Vertex.cpp:32]: Best not to do one-liners in switch, easy to miss

```suggestion
    case Qt::LeftButton: 
      handleLeftClickEvent(pEvent); 
      break;
    case Qt::RightButton: 
      emit rightClickPressed(mapToParent(pEvent->position())); 
      break;
```

File: src/visualizer/vertex/Vertex.cpp [hunk 16]
```diff
@@ -4,18 +4,61 @@
...
 }
 
-auto Vertex::mousePressEvent(QMouseEvent*) -> void { setPalette(Qt::red); }
-} // namespace age::visualizer
+constexpr auto Vertex::distanceToCenter(QPoint const& point) -> double {
+  auto distToCenter = WIDGET_SIZE / 2;
+  auto distEventToCenterX = distToCenter - point.x();
+  auto distEventToCenterY = distToCenter - point.y();
+  return std::sqrt(distEventToCenterX * distEventToCenterX + distEventToCenterY * distEventToCenterY);
+}
```
Comment (LoghinVladDev) [inline src/visualizer/vertex/Vertex.cpp]: EOL
```suggestion
}

```

File: src/visualizer/vertex/Vertex.cpp [hunk 17]
```diff
@@ -4,18 +4,61 @@
...
+  raise();
 }
 
-auto Vertex::mousePressEvent(QMouseEvent*) -> void { setPalette(Qt::red); }
-} // namespace age::visualizer
+constexpr auto Vertex::distanceToCenter(QPoint const& point) -> double {
+  auto distToCenter = WIDGET_SIZE / 2;
+  auto distEventToCenterX = distToCenter - point.x();
+  auto distEventToCenterY = distToCenter - point.y();
+  return std::sqrt(distEventToCenterX * distEventToCenterX + distEventToCenterY * distEventToCenterY);
```
Comment (LoghinVladDev) [inline src/visualizer/vertex/Vertex.cpp]: Minor preference
```suggestion
  auto distPointToCenterX = distToCenter - point.x();
  auto distPointToCenterY = distToCenter - point.y();
  return std::sqrt(distPointToCenterX * distPointToCenterX + distPointToCenterY * distPointToCenterY);
```

File: src/visualizer/vertex/Vertex.hpp [hunk 18]
```diff
@@ -3,22 +3,35 @@
...
+#include <cmath>
+#include <intern/QtDefines.hpp>
 
 namespace age::visualizer {
 class Vertex : public QWidget {
   Q_OBJECT
 public:
   explicit Vertex(int x, int y, QWidget* pParent) noexcept;
+  explicit Vertex(QPoint const& point, QWidget* pParent) noexcept;
+signals:
```
Comment (LoghinVladDev) [inline src/visualizer/vertex/Vertex.hpp]: NL between access specifiers. Even though `signals` is just `public`

```suggestion
  explicit Vertex(QPoint const& point, QWidget* pParent) noexcept;
  
signals:
```

File: test/unittests/qtMock/QAction [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+
+struct QAction : QObject {
+  explicit QAction(QObject* = nullptr) {}
+  explicit QAction(QString const&, QObject* = nullptr) {}
+  static auto exec() noexcept -> int {
+    return 0;
+  }
+  auto triggered() const {
+    // empty on purpose
+  }
```
Comment (LoghinVladDev) [inline test/unittests/qtMock/QAction]: These can be left empty, not analyzed
```suggestion
  auto triggered() const {}
```

File: test/unittests/qtMock/QMenu [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+//
+#pragma once
+
+#include <QWidget>
+
+struct QMenu : QWidget {
+  explicit QMenu(QWidget* = nullptr) {}
+  auto addAction(QAction*) const {
+    // empty on purpose
+  }
```
Comment (LoghinVladDev) [inline test/unittests/qtMock/QMenu]: Can be left empty
```suggestion
  auto addAction(QAction*) const {}
```

File: test/unittests/qtMock/QMenu [hunk 21]
```diff
@@ -0,0 +1,16 @@
...
+  auto exec(QPoint const&) const {
+    // empty on purpose
+  }
```
Comment (LoghinVladDev) [inline test/unittests/qtMock/QMenu]: Can be left empty
```suggestion
  auto exec(QPoint const&) const {}
```

File: test/unittests/qtMock/QObject [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+
+struct QObject {
+  explicit QObject(QObject* = nullptr) {}
+
+  template <
+      typename PointerToMemberFunction,
+      typename Functor
+  > static auto connect(QObject*, PointerToMemberFunction, Functor) {
+    // empty on purpose
+  }
```
Comment (LoghinVladDev) [inline test/unittests/qtMock/QObject]: Can be left empty
```suggestion
  > static auto connect(QObject*, PointerToMemberFunction, Functor) {}
```

File: test/unittests/qtMock/QObject [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+
+  template <
+      typename PointerToMemberFunction,
+      typename Functor
+  > static auto connect(QObject*, PointerToMemberFunction, QObject*, Functor) {
+    // empty on purpose
+  }
```
Comment (LoghinVladDev) [inline test/unittests/qtMock/QObject]: Can be left empty
```suggestion
  > static auto connect(QObject*, PointerToMemberFunction, QObject*, Functor) {}
```

File: test/unittests/qtMock/QPainter [hunk 24]
```diff
@@ -7,13 +7,18 @@
...
-  QPainter(QPaintDevice* pPaintDevice) {}
-  auto setPen(Qt::QColor) {}
-  auto drawEllipse(int, int, int, int) {}
+  explicit QPainter(QPaintDevice*) {}
+  auto setPen(Qt::QColor) const {
+    // empty on purpose
+  }
+  auto drawEllipse(int, int, int, int) const {
+    // empty on purpose
+  }
```
Comment (LoghinVladDev) [inline test/unittests/qtMock/QPainter]: Can be left empty
```suggestion
  auto drawEllipse(int, int, int, int) const {}
```

File: test/unittests/qtMock/QPainter [hunk 25]
```diff
@@ -7,13 +7,18 @@
...
 struct QPainter {
 
 public:
-  QPainter(QPaintDevice* pPaintDevice) {}
-  auto setPen(Qt::QColor) {}
-  auto drawEllipse(int, int, int, int) {}
+  explicit QPainter(QPaintDevice*) {}
+  auto setPen(Qt::QColor) const {
+    // empty on purpose
+  }
```
Comment (LoghinVladDev) [inline test/unittests/qtMock/QPainter]: Can be left empty
```suggestion
  auto setPen(Qt::QColor) const {
    // empty on purpose
  }
```

File: test/unittests/qtMock/QWidget [hunk 26]
```diff
@@ -3,25 +3,54 @@
...
 
 struct QPaintEvent;
 
-struct QWidget : QPaintDevice {
-  auto resize(int, int) noexcept -> void {}
-  auto show() noexcept -> void {}
+struct QWidget : QPaintDevice, QObject {
+  auto resize(int, int) const noexcept -> void {
+    // empty on purpose
+  }
```
Comment (LoghinVladDev) [inline test/unittests/qtMock/QWidget]: Can be left empty
```suggestion
  auto resize(int, int) const noexcept -> void {}
```

File: test/unittests/qtMock/QWidget [hunk 27]
```diff
@@ -3,25 +3,54 @@
...
+  auto show() const noexcept -> void {
+    // empty on purpose
+  }
```
Comment (LoghinVladDev) [inline test/unittests/qtMock/QWidget]: Can be left empty
```suggestion
  auto show() const noexcept -> void {}
```

File: test/unittests/qtMock/QWidget [hunk 28]
```diff
@@ -3,25 +3,54 @@
...
-  virtual auto mousePressEvent(QMouseEvent*) -> void {};
-  auto move(int, int) noexcept -> void {}
-  virtual auto paintEvent(QPaintEvent*) -> void {}
-  auto setPalette(QPalette const &) -> void {}
-  auto palette() -> QPalette const & { return _defaultPalette; }
-  auto backgroundRole() -> QPalette::ColorRole { return QPalette::_unused; }
-
+  virtual auto mousePressEvent(QMouseEvent*) -> void {
+      // empty on purpose
+  }
```
Comment (LoghinVladDev) [inline test/unittests/qtMock/QWidget]: Can be left empty
```suggestion
  virtual auto mousePressEvent(QMouseEvent*) -> void {}
```

File: test/unittests/qtMock/QWidget [hunk 29]
```diff
@@ -3,25 +3,54 @@
...
+  virtual auto mouseMoveEvent(QMouseEvent*) -> void {
+      // empty on purpose
+  }
```
Comment (LoghinVladDev) [inline test/unittests/qtMock/QWidget]: Can be left empty
```suggestion
  virtual auto mouseMoveEvent(QMouseEvent*) -> void {}
```

File: test/unittests/qtMock/QWidget [hunk 30]
```diff
@@ -3,25 +3,54 @@
...
+  virtual auto mouseReleaseEvent(QMouseEvent*) -> void {
+      // empty on purpose
+  }
```
Comment (LoghinVladDev) [inline test/unittests/qtMock/QWidget]: Can be left empty
```suggestion
  virtual auto mouseReleaseEvent(QMouseEvent*) -> void {}
```

File: test/unittests/qtMock/QWidget [hunk 31]
```diff
@@ -3,25 +3,54 @@
...
+  auto move(QPoint const&) const noexcept -> void {
+    // empty on purpose
+  }
```
Comment (LoghinVladDev) [inline test/unittests/qtMock/QWidget]: Can be left empty
```suggestion
  auto move(QPoint const&) const noexcept -> void {}
```

File: test/unittests/qtMock/QWidget [hunk 32]
```diff
@@ -3,25 +3,54 @@
...
+  auto move(int, int) const noexcept -> void {
+    // empty on purpose
+  }
```
Comment (LoghinVladDev) [inline test/unittests/qtMock/QWidget]: Can be left empty
```suggestion
  auto move(int, int) const noexcept -> void {}
```

File: test/unittests/qtMock/QWidget [hunk 33]
```diff
@@ -3,25 +3,54 @@
...
+  virtual auto paintEvent(QPaintEvent*) -> void {
+    // empty on purpose
+  }
```
Comment (LoghinVladDev) [inline test/unittests/qtMock/QWidget]: Can be left empty
```suggestion
  virtual auto paintEvent(QPaintEvent*) -> void {}
```

File: test/unittests/qtMock/QWidget [hunk 34]
```diff
@@ -3,25 +3,54 @@
...
+  auto setPalette(QPalette const &) const noexcept -> void {
+    // empty on purpose
+  }
```
Comment (LoghinVladDev) [inline test/unittests/qtMock/QWidget]: Can be left empty
```suggestion
  auto setPalette(QPalette const &) const noexcept -> void {}
```

File: test/unittests/qtMock/QWidget [hunk 35]
```diff
@@ -3,25 +3,54 @@
...
+  auto palette() const noexcept -> QPalette const & { return _defaultPalette; }
+  auto backgroundRole() const noexcept -> QPalette::ColorRole { return QPalette::_unused; }
+  auto mapToParent(QPointF const&) const -> QPointF { return {}; }
+  auto mapToParent(QPoint const&) const -> QPoint { return {}; }
+  auto mapToGlobal(QPoint const&) const -> QPoint { return {}; }
+  auto mapToGlobal(QPointF const&) const -> QPointF { return {}; }
+  auto raise() const {
+    // empty on purpose
+  }
```
Comment (LoghinVladDev) [inline test/unittests/qtMock/QWidget]: Can be left empty
```suggestion
  auto raise() const {}
```

**Review by LoghinVladDev — CHANGES_REQUESTED** [review]
Looks great! Only minor findings, should be good to go from suggetions

File: src/visualizer/vertex/Vertex.cpp (hunk 15 above)
Comment (StefanPaulet) [reply src/visualizer/vertex/Vertex.cpp:32]: Clang-format requires it to be a one-liner

**Review by LoghinVladDev — APPROVED** [review]
LGTM

**Pull request merged**